File: 1472-design-browser-history/1472-design-browser-history.cpp
class DoublyList{    
public:
    string data;
    DoublyList *prev, *next;
    DoublyList(string url){
        this->data = url;
        this->prev = this->next = NULL;
    }
};
class BrowserHistory {
private:
    DoublyList *root;
public:
    BrowserHistory(string homepage) {
        root = new DoublyList(homepage);
    }
    
    void visit(string url) {
        DoublyList *temp = new DoublyList(url);
        temp->prev = root;
        root->next = temp;
        root = temp;
    }
    
    string back(int steps) {
        while(root->prev != NULL && steps){
            root = root->prev;
            steps--;
        }
        return root->data;
    }
    
    string forward(int steps) {
        while(root->next != NULL && steps){
            root = root->next;
            steps--;
        }
        return root->data;
    }
};

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * BrowserHistory* obj = new BrowserHistory(homepage);
 * obj->visit(url);
 * string param_2 = obj->back(steps);
 * string param_3 = obj->forward(steps);
 */

File: 1472-design-browser-history/1472-design-browser-history.py
# class ListNode:
#     def __init__(self, val, prev=None, next=None):
#         self.val = val
#         self.prev = prev
#         self.next = next

# class BrowserHistory:
#     def __init__(self, homepage: str):
#         self.curr = ListNode(homepage)
#     # O(1)
#     def visit(self, url: str) -> None:
#         self.curr.next = ListNode(url, self.curr)
#         self.curr = self.curr.next
    
#     # O(n)
#     def back(self, steps: int) -> str:
#         while self.curr.prev and steps > 0:
#             self.curr = self.curr.prev
#             steps -= 1
#         return self.curr.val

#     # O(n)
#     def forward(self, steps: int) -> str:
#         while self.curr.next and steps > 0:
#             self.curr = self.curr.next
#             steps -= 1
#         return self.curr.val

class BrowserHistory:
    def __init__(self, homepage: str):
        self.i = 0
        self.len = 1
        self.history = [homepage]
        
    # O(1)
    def visit(self, url: str) -> None:
        if self.i + 1 < len(self.history):
            self.history[self.i + 1] = url
        else:
            self.history.append(url)
        self.i += 1
        self.len = self.i + 1
    
    # O(1)
    def back(self, steps: int) -> str:
        self.i = max(0, self.i - steps)
        return self.history[self.i]

    # O(1)
    def forward(self, steps: int) -> str:
        self.i = min(self.len - 1, self.i + steps)
        return self.history[self.i]

# Your BrowserHistory object will be instantiated and called as such:
# obj = BrowserHistory(homepage)
# obj.visit(url)
# param_2 = obj.back(steps)
# param_3 = obj.forward(steps)